File: src/org/nextras/orm/intellij/completion/EntityPropertiesProvider.java
package org.nextras.orm.intellij.completion;

import com.intellij.codeInsight.completion.CompletionParameters;
import com.intellij.codeInsight.completion.CompletionResultSet;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.jetbrains.php.lang.documentation.phpdoc.psi.tags.PhpDocPropertyTag;
import com.jetbrains.php.lang.parser.PhpElementTypes;
import com.jetbrains.php.lang.psi.elements.*;
import org.jetbrains.annotations.NotNull;
import org.nextras.orm.intellij.utils.OrmUtils;

import java.util.Arrays;
import java.util.Collection;
import java.util.stream.Collectors;
import java.util.stream.Stream;


public class EntityPropertiesProvider
{
	public void fillCompletionVariants(@NotNull CompletionParameters parameters, @NotNull CompletionResultSet result)
	{
		PsiElement element = parameters.getPosition().getParent();

		ParameterList parameterList = PsiTreeUtil.getParentOfType(element, ParameterList.class);
		if (parameterList == null) {
			return;
		}
		if (element.getParent() instanceof PhpPsiElement
			&& ((PhpPsiElement) element.getParent()).getPrevPsiSibling() != null
			&& ((PhpPsiElement) element.getParent()).getPrevPsiSibling().getNode().getElementType() == PhpElementTypes.ARRAY_KEY) {
			return;
		}
		PsiElement context = parameterList.getContext();
		if (!(context instanceof MethodReference)) {
			return;
		}
		MethodReference methodReference = (MethodReference) context;
		PsiElement methodEl = methodReference.resolve();
		if (methodEl == null || !(methodEl instanceof Method)) {
			return;
		}
		Method method = (Method) methodEl;
		if (!(method.getName().equals("findBy") || method.getName().equals("getBy") || method.getName().equals("orderBy"))) {
			return;
		}

		Project project = parameters.getEditor().getProject();
		if (project == null) {
			return;
		}


		String fieldExpression = parameters.getOriginalPosition().getText();
		String[] path = fieldExpression.split("->", -1);

		Collection<PhpClass> queriedEntities = OrmUtils.findQueriedEntities(methodReference, path);
		for (PhpClass cls : queriedEntities) {

			if (cls.getDocComment() == null) {
				continue;
			}
			for (PhpDocPropertyTag phpDocPropertyTag : cls.getDocComment().getPropertyTags()) {

				Stream<String> types = phpDocPropertyTag.getType().getTypesSorted().stream()
					.filter(s -> !s.contains("Nextras\\Orm\\Relationships") && !s.equals("?"))
					.map(s -> s.startsWith("\\") ? s.substring(1) : s);

				String strPath = String.join("->", Arrays.copyOfRange(path, 0, path.length - 1));
				if (strPath.length() > 0) {
					strPath += "->";
				}
				String fieldName = phpDocPropertyTag.getProperty().getText().substring(1);
				strPath += fieldName;

				result.addElement(LookupElementBuilder.create(strPath)
					.withPresentableText(fieldName)
					.withTypeText(types.collect(Collectors.joining("|"))));
			}
			if (path.length == 1) {
				result.addElement(LookupElementBuilder.create("this").withTypeText(cls.getType().toString()));
				result.addElement(LookupElementBuilder.create(cls.getFQN()).withTypeText(cls.getType().toString()));
			}
		}
	}
}

File: src/org/nextras/orm/intellij/completion/EntityPropertiesProvider.kt
package org.nextras.orm.intellij.completion

import com.intellij.codeInsight.completion.CompletionParameters
import com.intellij.codeInsight.completion.CompletionResultSet
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.psi.util.PsiTreeUtil
import com.jetbrains.php.PhpIcons
import com.jetbrains.php.lang.parser.PhpElementTypes
import com.jetbrains.php.lang.psi.elements.Method
import com.jetbrains.php.lang.psi.elements.MethodReference
import com.jetbrains.php.lang.psi.elements.ParameterList
import com.jetbrains.php.lang.psi.elements.PhpPsiElement
import org.nextras.orm.intellij.utils.OrmUtils

class EntityPropertiesProvider {
	private val relationshipRegexp = ".*\\{(?:m:m|m:1|1:m|1:1).+".toRegex()

	fun fillCompletionVariants(parameters: CompletionParameters, result: CompletionResultSet) {
		val element = parameters.position.parent

		val parameterList = PsiTreeUtil.getParentOfType(element, ParameterList::class.java) ?: return
		if ((element.parent as? PhpPsiElement)?.prevPsiSibling?.node?.elementType === PhpElementTypes.ARRAY_KEY) {
			return
		}

		val context = parameterList.context as? MethodReference ?: return
		val method = context.resolve() as? Method ?: return

		if (!(
				method.name == "findBy"
					|| method.name == "getBy"
					|| method.name == "getByChecked"
					|| method.name == "orderBy"
				)
		) {
			return
		}

		val project = parameters.editor.project ?: return
		val fieldExpression = parameters.originalPosition!!.text
		val fieldExpressionLen = parameters.editor.caretModel.offset - parameters.originalPosition!!.textOffset
		val expression = fieldExpression.substring(0, fieldExpressionLen)

		val isV3 = OrmUtils.isV3(project)
		val (sourceCls, path) = OrmUtils.parsePathExpression(expression, isV3)
		val classSuffix = when (isV3) {
			true -> "->"
			false -> "::"
		}

		val queriedEntities = OrmUtils.findQueriedEntities(context, sourceCls, path)
		queriedEntities
			.filter { it.docComment != null }
			.forEach { cls ->
				var strPathPrefix = (sourceCls?.let { it + classSuffix } ?: "") +
					path.dropLast(1).joinToString("->")
				if (path.size > 1) {
					strPathPrefix += "->"
				}

				for (phpDocPropertyTag in cls.docComment!!.propertyTags) {
					val types = phpDocPropertyTag.type.typesSorted
						.filter { !it.contains("Nextras\\Orm\\Relationships") && it != "?" }
						.map {
							when (it.startsWith("\\")) {
								true -> it.substring(1)
								false -> it
							}
						}

					val isRelationship = phpDocPropertyTag.text.matches(relationshipRegexp)
					val sep = if (isRelationship) "->" else ""
					val fieldName = phpDocPropertyTag.property!!.text.substring(1)
					val strPath = strPathPrefix + fieldName

					result.addElement(
						LookupElementBuilder.create(strPath + sep)
							.withPresentableText(fieldName)
							.withIcon(PhpIcons.FIELD_ICON)
							.withTypeText(types.joinToString("|"))
					)
				}

				if (isV3 && path.size == 1 && sourceCls == null) {
					result.addElement(
						LookupElementBuilder.create("this->")
							.withPresentableText("this")
							.withIcon(PhpIcons.CLASS_ICON)
							.withTypeText(cls.type.toString())
					)
				}
				if (path.size == 1 && sourceCls == null)
					result.addElement(
						LookupElementBuilder.create(cls.fqn + classSuffix)
							.withPresentableText(cls.fqn)
							.withIcon(PhpIcons.CLASS_ICON)
							.withTypeText(cls.type.toString())
					)
			}
	}
}
